File: src/main/java/com/almasb/zeph/ui/InventoryView.java
package com.almasb.zeph.ui;

import com.almasb.fxgl.texture.Texture;
import com.almasb.zeph.Config;
import com.almasb.zeph.ConfigKt;
import com.almasb.zeph.character.CharacterEntity;
import com.almasb.zeph.item.Armor;
import com.almasb.zeph.item.Item;
import com.almasb.zeph.item.UsableItem;
import com.almasb.zeph.item.Weapon;
import javafx.beans.binding.Bindings;
import javafx.beans.binding.BooleanBinding;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.collections.ListChangeListener;
import javafx.geometry.Point2D;
import javafx.geometry.Pos;
import javafx.scene.Cursor;
import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.control.Tooltip;
import javafx.scene.input.MouseButton;
import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.Region;
import javafx.scene.layout.StackPane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Circle;
import javafx.scene.shape.Rectangle;
import javafx.scene.shape.Shape;
import javafx.scene.text.Font;
import javafx.scene.text.Text;
import javafx.util.Duration;
import javafx.util.Pair;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import static com.almasb.fxgl.dsl.FXGL.*;

public class InventoryView extends Parent {

    private static final int BG_WIDTH = 200;
    private static final int BG_HEIGHT = 240;

    private boolean isMinimized = false;

    // K - index, V - if free? TODO: double check
    private Map<Integer, Boolean> slots = new HashMap<>();

    private Group itemGroup = new Group();

    private TooltipView tooltip = ConfigKt.getUITooltip();
    private BooleanBinding isTooltipVisible;

    private ListChangeListener<Item> listener;

    private CharacterEntity player;

    public InventoryView(CharacterEntity player) {
        relocate(getAppWidth() - 200, getAppHeight() - 240);

        var btn = new Rectangle(30, 20);
        btn.setStroke(Color.WHITE);
        btn.relocate(BG_WIDTH - 46, -22);
        btn.setOnMouseClicked(e -> {

            if (isReleaseMode()) {
                play("ui_slide.wav");
            }

            animationBuilder()
                    .duration(Duration.seconds(0.33))
                    .translate(this)
                    .from(isMinimized ? new Point2D(0.0, BG_HEIGHT) : new Point2D(0.0, 0.0))
                    .to(isMinimized ? new Point2D(0.0, 0.0) : new Point2D(0.0, BG_HEIGHT))
                    .buildAndPlay();

            isMinimized = !isMinimized;
        });

        Rectangle border = new Rectangle(BG_WIDTH * 2 + 3, BG_HEIGHT + 5);
        border.setStrokeWidth(2);
        border.setArcWidth(10);
        border.setArcHeight(10);

        Shape borderShape = Shape.union(border, new Circle(BG_WIDTH*2 + 3 - 30, 0.0, 30));
        borderShape.setFill(Color.rgb(25, 25, 25, 0.8));
        borderShape.setStroke(Color.WHITE);
        borderShape.setTranslateX(-BG_WIDTH - 3);

        Texture background = getAssetLoader().loadTexture("ui/inventory_right.png");

        EquipmentView equipView = new EquipmentView(player);
        equipView.setTranslateX(-BG_WIDTH);

        getChildren().addAll(borderShape, background, equipView, itemGroup, btn);

        this.player = player;

        for (int i = 0; i < 30; i++) {
            slots.put(i, true);
        }

        listener = new ListChangeListener<Item>() {
            @Override
            public void onChanged(ListChangeListener.Change<? extends Item> change) {
                while (change.next()) {
                    if (change.wasAdded()) {
                        for (Item item : change.getAddedSubList()) {
                            addItem(item);
                        }
                    }
                    else if (change.wasRemoved()) {
                        for (Item item : change.getRemoved()) {

                            for (Iterator<Node> it = itemGroup.getChildren().iterator(); it.hasNext(); ) {
                                Node node = it.next();

                                if (node.getUserData() != null) {
                                    Pair<Item, Integer> data = (Pair<Item, Integer>) node.getUserData();

                                    if (data.getKey() == item) {
                                        slots.put(data.getValue(), true);
                                        it.remove();
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        };

        player.getInventory().itemsProperty().forEach(this::addItem);

        player.getInventory().itemsProperty().addListener(listener);

        // TODO: should have 1 tooltip object for everything in the game
        // TODO: description should have TextFlow, so we can color some text, e.g. 10% damage, runes, etc.

        isTooltipVisible = Bindings.createBooleanBinding(() -> false, player.getInventory().itemsProperty());

        // TODO: and NEWly added items
        itemGroup.getChildren().forEach(node -> {
            isTooltipVisible = isTooltipVisible.or(node.hoverProperty());
        });


        //tooltip.visibleProperty().bind(isTooltipVisible);

    }

    private int getNextFreeSlot() {
        for (int i = 0; i < 30; i++) {
            if (slots.get(i))
                return i;
        }

        return -1;
    }

    private void addItem(Item item) {
        int index = getNextFreeSlot();
        slots.put(index, false);

        StackPane view = new StackPane(texture(item.getDescription().getTextureName()));
        view.setAlignment(Pos.BOTTOM_RIGHT);
        view.setUserData(new Pair<>(item, index));
        view.setTranslateX((index % 5) * 40 + 3);
        view.setTranslateY((index / 5) * 40 + 3);
        view.setPickOnBounds(true);

        view.hoverProperty().addListener((o, old, isHover) -> {
            if (isHover) {
                tooltip.setItem(item);
            }
        });

        view.setOnMouseClicked(event -> {

            if (event.getButton() == MouseButton.PRIMARY) {
                if (item instanceof Weapon) {
                    player.getPlayerComponent().equipWeapon((Weapon) item);
                } else if (item instanceof Armor) {
                    player.getPlayerComponent().equipArmor((Armor) item);
                } else if (item instanceof UsableItem) {
                    player.getCharacterComponent().useItem((UsableItem) item);
                }

                // TODO: other usable types
            }
        });
        view.setCursor(Cursor.HAND);

        var text = getUIFactoryService().newText("", Color.WHITE, 12.0);
        text.textProperty().bind(player.getInventory().getData(item).get().quantityProperty().asString());
        text.setStrokeWidth(1.5);

        view.getChildren().addAll(text);

        itemGroup.getChildren().add(view);
    }
}

File: src/main/kotlin/com/almasb/zeph/ui/InventoryView.kt
package com.almasb.zeph.ui

import com.almasb.fxgl.dsl.FXGL
import com.almasb.fxgl.dsl.getUIFactoryService
import com.almasb.fxgl.dsl.image
import com.almasb.fxgl.inventory.Inventory
import com.almasb.fxgl.inventory.ItemStack
import com.almasb.fxgl.ui.FXGLScrollPane
import com.almasb.zeph.Gameplay
import com.almasb.zeph.MouseGestures
import com.almasb.zeph.data.Data
import com.almasb.zeph.item.Item
import com.almasb.zeph.item.UsableItem
import com.almasb.zeph.item.Weapon
import javafx.collections.ListChangeListener
import javafx.geometry.Point2D
import javafx.geometry.Pos
import javafx.scene.Cursor
import javafx.scene.Group
import javafx.scene.Parent
import javafx.scene.control.ScrollPane
import javafx.scene.input.MouseButton
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.scene.shape.Rectangle
import javafx.scene.text.Font

/**
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
class InventoryView(val inventory: Inventory<Item>,

                    /**
                     * Width and height in number of cells.
                     */
                    val widthInCells: Int, val heightInCells: Int) : Parent() {

    private val contentPane = Pane()

    private val scrollPane = FXGLScrollPane(contentPane)

    private val itemGroup = Group()
    private val listener: ListChangeListener<ItemStack<Item>>

    private val itemPanes
        get() = itemGroup.children.map { it as ItemPane }

    private val mouseGestures = MouseGestures(itemGroup)

    var onItemClicked: (Item) -> Unit = {}

    val layoutWidth: Double
        get() = scrollPane.prefWidth

    val layoutHeight: Double
        get() = scrollPane.prefHeight

    init {
        val bg = Rectangle(40.0 * widthInCells, inventory.capacity / widthInCells * 40.0)
        bg.arcWidth = 15.0
        bg.arcHeight = 15.0
        bg.stroke = Color.AQUA
        bg.strokeWidth = 1.5


        scrollPane.setPrefSize(bg.width + bg.strokeWidth * 2 + 10.0, 40.0 * heightInCells)
        scrollPane.hbarPolicy = ScrollPane.ScrollBarPolicy.NEVER

        contentPane.background = Background(BackgroundImage(image("ui/item_frame.png"),
                BackgroundRepeat.REPEAT, BackgroundRepeat.REPEAT, null, null))

        children += scrollPane

        contentPane.setPrefSize(bg.width, bg.height)
        contentPane.maxHeight = bg.height

        // populate item panes
        for (y in 0 until inventory.capacity / widthInCells) {
            for (x in 0 until widthInCells) {
                val pane = ItemPane()
                pane.translateX = x * 40 + 3.toDouble()
                pane.translateY = y * 40 + 3.toDouble()

                mouseGestures.makeDraggable(pane) {
                    swap(pane)
                }

                pane.setOnMouseClicked {
                    if (pane.isEmpty)
                        return@setOnMouseClicked

                    if (it.button == MouseButton.PRIMARY && it.clickCount == 2) {
                        val item = pane.itemStack!!.userItem
                        onItemClicked(item)
                    }
                }

                itemGroup.children += pane
            }
        }

        contentPane.children += itemGroup

        listener = ListChangeListener { change ->
            while (change.next()) {
                if (change.wasAdded()) {
                    for (item in change.addedSubList) {
                        addItem(item)
                    }
                } else if (change.wasRemoved()) {
                    for (stack in change.removed) {
                        itemPanes.filter { it.itemStack === stack}
                                .forEach {
                                    it.itemStack = null
                                }
                    }
                }
            }
        }

        inventory.itemsProperty().forEach { addItem(it) }
        inventory.itemsProperty().addListener(listener)
        
        // TEST

        inventory.add(UsableItem(Data.UsableItems.MANA_POTION))
        inventory.add(UsableItem(Data.UsableItems.HEALING_POTION))

        inventory.add(UsableItem(Data.UsableItems.TELEPORTATION_STONE))
        inventory.add(UsableItem(Data.UsableItems.TELEPORTATION_STONE))
        inventory.add(Weapon(Data.Weapons.OneHandedSwords.GUARD_SWORD))
    }

    private fun swap(pane: ItemPane) {
        val closestPane = itemPanes
                .sortedBy {
                    Point2D(it.translateX, it.translateY)
                            .distance(Point2D(pane.layoutX + pane.translateX, pane.layoutY + pane.translateY))
                }
                .first()

        pane.relocate(0.0, 0.0)

        if (closestPane === pane)
            return

        // TODO: implement proper swap, now we need to set to null to clean up
        val stack = pane.itemStack

        pane.itemStack = null
        pane.itemStack = closestPane.itemStack
        closestPane.itemStack = null
        closestPane.itemStack = stack
    }

    private fun getNextFreeSlot(): ItemPane? {
        return itemPanes.find { it.isEmpty }
    }

    private fun addItem(stack: ItemStack<Item>) {
        getNextFreeSlot()?.let {
            it.itemStack = stack
        }
    }
}

private class ItemPane : StackPane() {

    private val text = FXGL.getUIFactoryService().newText("", Color.WHITE, 12.0)

    var itemStack: ItemStack<Item>? = null
        set(value) {
            if (value == null) {

                if (itemStack != null) {
                    children.removeAt(0)

                    text.textProperty().unbind()
                    text.text = ""
                }

            } else {
                val view = FXGL.texture(value.userItem.description.textureName)

                children.add(0, view)

                text.textProperty().bind(value.quantityProperty().asString())

                // TODO: remove when value == null?
                setOnTooltipHover { t: TooltipView ->
                    t.setItem(value.userItem)
                }
            }

            field = value
        }

    val isEmpty: Boolean
        get() = itemStack == null

    init {
        alignment = Pos.BOTTOM_RIGHT
        isPickOnBounds = true
        cursor = Cursor.HAND

        text.strokeWidth = 1.5

        children += text
    }
}

class StorageView(val inventory: Inventory<Item>) : Parent() {
    val inventoryView = InventoryView(inventory, 20, 4)

    val closeButton = getUIFactoryService().newButton("Close")

    init {
        closeButton.fontProperty().unbind()
        closeButton.font = Font.font(11.0)
        closeButton.setPrefSize(213.0, 13.0)
        closeButton.setOnAction {
            Gameplay.closeStorage()
        }

        val vbox = VBox(inventoryView, closeButton)
        vbox.alignment = Pos.TOP_CENTER

        children += vbox
    }
}